commit 1b454ec8d2b09f5b8cd04e69a36527d7ad774c4b
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2019-06-07T08:54:56-07:00 (4 months ago)
    
    TRITON-1727 Add vmapi default user migration allowed config setting
    Reviewed by: Pedro Palazón Candel <pedro@joyent.com>
    Approved by: Pedro Palazón Candel <pedro@joyent.com>

diff --git a/docs/index.md b/docs/index.md
index 2073351..afeb83e 100644
--- a/docs/index.md
+++ b/docs/index.md
@@ -2267,6 +2267,11 @@ For any migration action (e.g. begin, sync, switch or abort) you can use the
 migration watch endpoint to show progress information for the running migration
 action.
 
+User migrations are disabled by default, but that can be changed globally by
+setting the VMAPI/SAPI user_migration_allowed setting to true, or
+changed on a per-instance basis by updating
+vm.internal_metadata.user_migration_allowed to true.
+
 ## VmMigrateList (GET /migrations)
 
 Returns the list of [Migration Objects](#migration-objects).
@@ -2662,6 +2667,7 @@ production.
 | **experimental_fluentd_host**  | String |                                                                              |
 | **docker_tag_re**              | String | Tags matching regex are treated with Docker tag semantics                    |
 | **migration_send_mbps_limit**  | Number | Limit of transfer rate, in megabits per second, for individual migrations    |
+| **user_migration_allowed**     | Boolean | Whether user migrations are allowed (default is false). This setting can also be overruled on a per-instance basis through the vm.internal_metadata.user_migration_allowed setting. |
 
 `docker_tag_re` must be a valid regular expression string -- more concretely,
 what Javascript's RegExp() considers valid. Docker tags can be added during
diff --git a/lib/vm-migration/migrate.js b/lib/vm-migration/migrate.js
index 0f50e11..c51a747 100644
--- a/lib/vm-migration/migrate.js
+++ b/lib/vm-migration/migrate.js
@@ -43,6 +43,10 @@ function validateMigrationBegin(vm, ctx, callback) {
     assert.object(vm, 'vm');
     assert.optionalObject(vm.tags, 'vm.tags');
     assert.object(ctx, 'ctx');
+    assert.object(ctx.app, 'ctx.app');
+    assert.object(ctx.app.options, 'ctx.app.options');
+    assert.optionalBool(ctx.app.options.userMigrationAllowed,
+        'ctx.app.options.userMigrationAllowed');
     assert.optionalObject(ctx.migrationRecord, 'ctx.migrationRecord');
     assert.func(callback, 'callback');
     assert.optionalUuid(ctx.ownerUuid, 'ctx.ownerUuid');
@@ -50,9 +54,14 @@ function validateMigrationBegin(vm, ctx, callback) {
     var vmAllowed = false;
 
     if (!ctx.ownerUuid) {
+        // This is an operator initiated migration - always allow it.
         vmAllowed = true;
     } else {
-        vmAllowed = vm.internal_metadata.user_migration_allowed === true;
+        // This is a user initiated migration - check if that is allowed.
+        vmAllowed = (ctx.app.options.userMigrationAllowed === true);
+        if (vm.internal_metadata.hasOwnProperty('user_migration_allowed')) {
+            vmAllowed = (vm.internal_metadata.user_migration_allowed === true);
+        }
     }
 
     if (!vmAllowed) {
@@ -302,6 +311,7 @@ function migrateVm(req, res, next) {
             }
 
             if (action === 'begin' || action === 'estimate') {
+                ctx.app = req.app;
                 validateMigrationBegin(vm, ctx, cb);
                 return;
             }
diff --git a/lib/vmapi.js b/lib/vmapi.js
index 6db1a56..41b92a7 100644
--- a/lib/vmapi.js
+++ b/lib/vmapi.js
@@ -47,6 +47,8 @@ function VmapiApp(options) {
     assert.object(options, 'options');
     assert.object(options.metricsManager, 'options.metricsManager');
     assert.optionalObject(options.log, 'options.log');
+    assert.optionalBool(options.userMigrationAllowed,
+        'options.userMigrationAllowed');
 
     // Fabric options
     assert.optionalObject(options.overlay, 'options.overlay');
diff --git a/package.json b/package.json
index d4ad1b3..124d9ff 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "vmapi",
   "description": "VMs API",
-  "version": "9.8.8",
+  "version": "9.8.9",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
diff --git a/sapi_manifests/vmapi/template b/sapi_manifests/vmapi/template
index 72e2630..b8b3372 100644
--- a/sapi_manifests/vmapi/template
+++ b/sapi_manifests/vmapi/template
@@ -27,6 +27,9 @@
         "enabled": false
 {{/fabric_cfg}}
     },
+{{#user_migration_allowed}}
+    "user_migration_allowed": {{user_migration_allowed}},
+{{/user_migration_allowed}}
 {{#migration_send_mbps_limit}}
     "migration_send_mbps_limit": {{migration_send_mbps_limit}},
 {{/migration_send_mbps_limit}}
diff --git a/server.js b/server.js
index e28a1a3..7fcc58f 100644
--- a/server.js
+++ b/server.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2018, Joyent, Inc.
+ * Copyright (c) 2019, Joyent, Inc.
  */
 
 /*
@@ -303,6 +303,7 @@ function startVmapiService() {
                 metricsManager: metricsManager,
                 moray: moray,
                 morayBucketsInitializer: morayBucketsInitializer,
+                userMigrationAllowed: config.user_migration_allowed,
                 zfs_send_mbps_limit: config.migration_send_mbps_limit,
                 overlay: config.overlay,
                 cnapi: config.cnapi, // TODO: TRITON-1295
